File: HW3/Job_Scheduling.py
# File Name : Job_Scheduling.py
# Description : 알고리즘 HW3 과제 중 7. Job Scheduling algorithm 구현 입니다. 
# input: Chapter 4-2, 27p에 있는 t1-t7


# Job List t1~t7
jobs = [
    (7, 8), (3, 7), (1, 5), (5, 9), (0, 2), (6, 8), (1, 6)
]

# sorting by starting time, jobs = (job_id, (start, end))
jobs = sorted(enumerate(jobs, start=1), key=lambda x: x[1][0])

# list for machine's job Schedule
machines = []

# 작업을 각 기계에 배정
for job_id, (start, end) in jobs:
    assigned = False
    # 각 기계를 순회하면서 현재 작업을 배정할 수 있는지 확인
    for machine in machines:
        # 기계의 마지막 작업 종료 시간보다 현재 작업 시작 시간이 늦으면 배정
        if machine[-1][1] <= start:
            machine.append((job_id, start, end))
            assigned = True
            break
    # 현재 작업을 배정할 수 있는 기계가 없다면, 새 기계에 작업을 추가
    if not assigned:
        machines.append([(job_id, start, end)])

# 시간표 생성
time_table = [[" " for _ in range(10)] for _ in range(len(machines))]

# 각 기계의 시간표에 작업 이름을 할당
for i, machine in enumerate(machines):
    for job_id, start, end in machine:
        for t in range(start, end):
            time_table[i][t] = f"t{job_id}"

# 시간표 출력
print("time ", end="")
for t in range(10):
    print(f"{t:>3}", end="")
print()

for i, row in enumerate(time_table):
    print(f"Machine {len(machines) - i}", end=" ")
    for cell in row:
        print(f"{cell:>3}", end="")
    print()

File: HW6/JobScheduling.java
// File Name : JobScheduling.java
// Description : 알고리즘 HW6 과제 중 4. Job Scheduling Algorithm 구현 입니다.
// - Input : 강의 노트 Chapter 8-2, 5p에 나와 있는 값 - m = 4, operation_time = [5, 2, 4, 3, 4, 7, 9, 2, 4, 1]
// - Output : Result Table
import java.util.*;

public class JobScheduling {

    static class Job {
        int jobId;    // Job(task) ID (t1, t2, ...)
        int jobTime;  // Time for Job

        public Job(int jobId, int jobTime) {
            this.jobId = jobId;
            this.jobTime = jobTime;
        }
    }

    // Scheduling Func
    public static void scheduleJobs(int m, int[] operationTime) {
        int n = operationTime.length;


        Job[] jobs = new Job[n];
        for (int i = 0; i < n; i++) {
            jobs[i] = new Job(i + 1, operationTime[i]);
        }

        // End time
        int[] machineEndTime = new int[m];
        Arrays.fill(machineEndTime, 0);

        // Scheduling table
        int[][] schedule = new int[m][13];
        for (int i = 0; i < m; i++) {
            Arrays.fill(schedule[i], 0);
        }

        // Find Early end
        for (Job job : jobs) {
            int minEndTimeMachine = 0;
            for (int i = 1; i < m; i++) {
                if (machineEndTime[i] < machineEndTime[minEndTimeMachine]) {
                    minEndTimeMachine = i;
                }
            }

            int startTime = machineEndTime[minEndTimeMachine];
            int endTime = startTime + job.jobTime;

            for (int t = startTime; t < endTime && t < 13; t++) {
                schedule[minEndTimeMachine][t] = job.jobId; 
            }

            // Update
            machineEndTime[minEndTimeMachine] = endTime;
        }

        // Print Result
        System.out.print("   m/t");
        for (int t = 0; t <= 12; t++) {
            System.out.printf("%4d", t);
        }
        System.out.println();


        for (int i = 0; i < m; i++) {
            System.out.printf("     %-2d", (i + 1));  
            for (int t = 0; t <= 12; t++) {
                if (schedule[i][t] != 0) {
                    System.out.printf(" t%-2d", schedule[i][t]);  
                } else {
                    System.out.printf("   "); 
                }
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        // Input
        int[] operationTime = {5, 2, 4, 3, 4, 7, 9, 2, 4, 1};
        
        int m = 4;

        // Func Call 
        scheduleJobs(m, operationTime);
    }
}
